### ISSUE TITLE
Report overall significance level? 

### ISSUE THREAD

**Issue by dmcglinn** [MoBiodiv/mobr#62]
John requested that practitioners would like to know an overall significance levels for each stage of the analysis. I'm a bit resistant to this because it de-emphasizes that the answer depends on the scale of the analysis but I can still see the importance. There is a precedent for an overall significance test from the spatial correlogram analyses. Legendre and Fortin (1989) suggest that overall significance can be assessed if there is at least one point that is significant at the level of alpha_prime = alpha / n where alpha is typically 0.05 and n is the number of tests which in our case will be number points along our rarefaction curves we examine. This is the Bonferroni method of correcting for multiple tests.

**Comment by dmcglinn** [comment]
Thanks @FelixMay  for providing this paper [Loosmore2006.pdf](https://github.com/MoBiodiv/mobr/files/375640/Loosmore2006.pdf) which does point out that the Bonferroni approach isn't relevant because of the non-independence of each of the tests. They suggest instead a sum of squared differences between the observed and the mean of the null permutations (eq 3) then compare that to the sum of squared differences of each null result relative to the null mean. This will give you a distribution of n permutation values which you can then evaluate how many times is the observed sum larger than the null sums. This sounds reasonable. Any thoughts here?

**Comment by rueuntal** [comment]
Sounds reasonable to me. I'm with you that only providing an overall
significance level goes against our goal, but would have no problem if we
provide both the overall significance and our plot showing how it changes
with scale. :)

**Comment by rueuntal** [comment]
Hi @dmcglinn @FelixMay @ngotelli  - Thanks Felix for pointing us to the Loosemore paper! Having read through it I think it provides insights for two things: 1. procedure to obtain an overall significance level for each curve, and 2. the implication for our current implementation. My comment will be a long one so I'll break it into these two pieces. Felix could you please let me know if my understanding is correct? Dan & Nick, it would be great to have your input too if you have the time to go over the paper but if not that's ok.

**Comment by rueuntal** [comment]
I'll start with 2. implementation for our current algorithm. I think the authors brought up an important point that could potentially change our interpretation of the current result. Just as a reminder, our algorithm builds a confidence interval around the effect of each factor, by simulating >100 times from a null model, and report the "significance level" at each point of evaluation along the curve. For example, if we evaluate the effect of aggregation at each plot level, there will be as many evaluations as the number of plots per treatment. Loosemore & Ford correctly pointed out that having one significant point is not equivalent as being significantly different from the null because of multi-tests. E.g., if all evaluations are independent (which is not the case), the probability of having all n points within the 95% hull would be .95^n, which is significantly lower than .95. 

I don't think we ever claim significance of the whole curve based on these results, so it is less of an issue. However, I think our interpretation implies that the effect is most drastic where the curve is outside of the hull. Loosemore & Ford also warned against this interpretation - because the curves are cumulative, effect at every scale not only represents the single-scale effect but also encompasses effect from all smaller scales. I tend to agree with them on this. What do other folks think?

**Comment by rueuntal** [comment]
If we all share the concern, I'm thinking that we can tweak our approach using Loosemore & Ford's overall significance measure. Namely, let's ditch reporting significance at each level - the magnitudes of the effects are accurate, it's only the significance that might be off. We can still plot the confidence hull but should make clear that one cannot directly reads the significance off the graph. Then instead of one p-value at each scale, we'd report an overall significance, calculated still using the simulations but across the entire range. 

I was thinking that maybe @FelixMay and I can resolve this by ourselves, but while I'm typing this seems increasingly to be a relatively major change, and it would be great if we can all agree. So @dmcglinn @ngotelli I'd love to hear comments from you too! Please let me know if anything needs to be clarified. Once we are all on the same page, I'll go ahead to tweak the code.

**Comment by dmcglinn** [comment]
Hey @rueuntal thanks for bringing this topic back up and being willing to tackle it. I think this is the best way to go if we are to going to report significance. I did find these resources: 

https://www.rdocumentation.org/packages/spatstat/versions/1.49-0/topics/dclf.test
and see FAQ 5 here: 
https://spatstat.github.io/FAQ.html

So it appears that it is already coded up in R for point pattern analysis. It is possible that their code could be generalized to work for our methods but this may end up being more difficult than actually just programming from scratch. I just thought I'd mention it though so you could at least double check the method.

**Comment by rueuntal** [comment]
Thanks for the reference @dmcglinn ! I wonder if there's one thing that R doesn't already have ;)

It's my impression too that it might be easier just coding the algorithm ourselves instead of tweaking others'.  If we are all good, I'll begin working on it this coming week.

**Comment by rueuntal** [comment]
I talked to Brian, and he echoed @dmcglinn 's earlier hesitation regarding the limitation of an overall p-value. Quoting his words - "The irony is that if you get an 'overall significance' you are now ignoring scale dependence. It is very possible to be significant at some scales and not at others. I would think twice before producing a p-value that has a very limited and specific interpretation."

So the question is do we want to provide a metric that's not technically incorrect, but may be difficult to interpret and have the risk to be misused. Dan, what do you think?

**Comment by dmcglinn** [comment]
Thanks for sharing Brian's input! An additional argument against an overall p-value is that we already provide the univariate randomization tests which accomplish in principle a similar goal although not after removing the effects of different drivers. If we provide folks with the 95% CI on the null models they are going to interpret divergence as statistical significance despite us saying not to and citing Loosmore. I think they will make this mistake regardless of whether or not we provide them with an overall significance test. Therefore it would appear the only advantage of providing an overall significance test is that it satisfies a "p-hacking" (http://www.nature.com/news/scientific-method-statistical-errors-1.14700) approach to ecology which would be satisfying to some users but we are obligated not to promote. **So given Brian's warnings I'm leaning away from providing the overall p-value.** I emailed Jon for his input and I'll post his response here if he sends one.

**Comment by ngotelli** [comment]
Hi all:


Sorry to be out of touch on this. Whether the p-values are valid or not, at the end of the day researchers always want a yes-no answer to tests for pattern detection. I myself am a big fan of this, even though I appreciate the problems that can arise with simplistic hypothesis testing. The best we can do is provide tools for the kinds of things researchers want to do. For that reason, I would be inclined to provide an overall p-value test, but also to give a 95% confidence interval. I am interested in giving scientists useful tools for the questions they are asking, not trying to police people into a particular form of analysis or interpretation. That being said, I am perfectly happy to go along with a decision to avoid providing a p-value. I just worry it will make the test less popular with users, who will inevitably have to answer to reviewers and editors when they make claims about patterns in their data.


Best,


Nick


________________________________

**Comment by rueuntal** [comment]
Thanks for the input, @ngotelli ! If I have to choose for my own research I wouldn't report an overall p-value, but I can see Nick's point of not policing what other researchers do, as long as it's done correctly. A compromise (which may be a bit over-the-top) that we can do is to provide the option of an overall p-value in the test but set the default to `FALSE`. If the user chooses to turn it on, it will come with a warning that the p-value returned is scale-insensitive and the user should be use caution in interpretation.

**Comment by ngotelli** [comment]
I think this is a good compromise. The default values guide users for the suggested settings and interpretations, but we give the options for standard p-values if users want them.


N.


________________________________

**Comment by FelixMay** [comment]
Dear mobr-team,
as Nick I like the compromise that Xiao suggested! It is a while ago that I read the Loosmore paper, but as far as I remember it should not be that difficult to implement their approach for our case. I also guess that it is easier to code it ourselves than adapting the spatstat::dclf.test function. When necessary, we can copy-paste and adapt their source code. @rueuntal let me know when you need help with the implementation, but I guess you will be faster than me, because you much better know the mobr data structures

**Comment by rueuntal** [comment]
Thanks @ngotelli & @FelixMay for the feedback! I'll start working on this tomorrow and keep you posted.

**Comment by jmchase** [comment]
Hi all, I got Dan's email about contributing to this discussion...but it looks like you have made some good headway already. Like others, I like Xiao's solution to have it there, with caveats.  I agree with Nick that people want to have a p-value. But I also agree with Brian that in this case, it is clear that p-values are strongly scale-sensitive, and this is our whole point, to some degree.  In all, I think people who are obsessed with p-values can get much of what they want with the randomization tests of Schao, PIE, etc.  One thought here is that it would be nice to have significance levels for both the smallest (alpha) and largest (gamma) scale.  Are we doing that now with the randomization tests, or is the largest (gamma) value still just a number?  Regardless, I think that if we give people p-values at local and regional scales, we can be a bit more circumscript in the continuous version.  So, again, I agree with the opinion that we should give them, but with caution.

**Comment by rueuntal** [comment]
Hi @jmchase - Thanks for chiming in! It seems that we are all on the same page now. I'll proceed accordingly. And looking at the code for the two-scale analysis it seems that we are reporting p-values at both alpha and gamma scales ( @dmcglinn & @FelixMay please correct me if I'm wrong).

**Comment by FelixMay** [comment]
Hi Xiao, Jon and Dan,
as far as I know there are no p-values at the "gamma" scale (treatment-scale) and there is no change to get p-values, because there is no replication at this scale, because there is jsut one (pooled) community for every treatment level. Is that right?

**Comment by jmchase** [comment]
Yes, this is exactly what I was wondering about.  I agree that the gamma scale has no replications, but now that we're using randomization tests, can't we get errors?  Or even a jackknife sort of thing?  I think people will want to have p-values at multiple scales....and it's always been annoying that we can't do this at the largest scale. Question is, do we try to help with this via some randomizations or such?  Or do we blow it off?  I'd like to try to find a way to get p-values at the regional scale if possible.

**Comment by rueuntal** [comment]
Hi @FelixMay - it has been a while since our discussion so it's getting a bit fuzzy in my mind. I'm looking at lines 153 - 161 in `mobr_boxplot.R`, and my read is that the p-values at the gamma-level (as well as the alpha-level) are calculated by randomly shuffling the treatment labels. Is that right?

**Comment by FelixMay** [comment]
@rueuntal - The shuffling you refer to is for the samples (alpha scale). If I remember correctly we discussed this issue sometime ago together with Nick, but I did not find the respective issue. For the treatment scale there are just two data points AFTER pooling. So there is nothing to shuffle AFTER pooling all data for the treatments. We can shuffle BEFORE pooling. Then we will get a distribution of biodiversity indices for the null assumption that there is no difference between the groups. We can get the rank of each treatment within this null distribution. In my understanding this gives us a test that each treatment differs from the overall mean, but this is something else then the difference between treatments. 
In short, I have no clear idea how to get a p-value at the gamma scale. When you have one let me know and I will be happy to implement it :o)

**Comment by rueuntal** [comment]
@FelixMay - You are right, if we are to shuffle at the gamma-scale there would only be two treatments and really nothing to shuffle with. The way it's currently implemented, I think, is to still do shuffling at the plot-level, and get the "shuffled" gamma-scale measurements from the reshuffled (and relabeled) plots. Does that sound reasonable?

**Comment by jmchase** [comment]
Hey all,  I'm not quite sure what Xiao is suggesting in the last message, but I think this makes sense.  Maybe I am being naïve, but I think it would be really useful if we can find a way to test the hypothesis that the two treatments differ at the gamma scale.  Of course, I've been doing this for years, and never figured it out before.  But it's always been a bit dissatisfying to me, reviewers, readers, etc.  So, is there any way we can think that can give us some estimate of significance between treatments at a larger scale in addition to the smaller scale.  Obviously, I understand why there are just two data points after pooling.  If Xiao's implementation above doesn't do this, couldn't we do some sort of jackknife of the data to get variance with N-1 (where N=gamma).  Or some randomization pulling a subset of the total gamma to get an estimate of variance around a second scale (that is, rather than having gamma be the whole set of observations, have it be a subset, so that we can still get estimates of two scales...or even more...but also get estimates of variance around these?  Again, as Nick has suggested, and I've heard from the empirical side...we need some sort of effect size and significance test...and if we can get it at multiple scales, we can get at much of what we want (i.e., bigger effects at smaller scales than bigger scales, or whatever) without having to try to put a single overall p-value on the whole test (which I agree with Brian, sort of defeats the purpose of our endeavor).  Make sense?

**Comment by FelixMay** [comment]
Hello again,
@rueuntal: Yes we could get the gamma-scale measurements after shuffling treatment labels. As I mentioned before this will give us a distribution for the overall mean, I think. Do you have any idea, how we can convert this into a relevant effect size and p-value? I am not sure

@jmchase: My best idea at the moment is to derive bootstrapped confidence intervals by drawing random samples WITH replacement and then do the pooling. However, I am not sure if this is statistically appropriate.

@ngotelli: Nick, are you reading this? Would be great to her your opinion on this issue?

**Comment by rueuntal** [comment]
@FelixMay - Ah, I see. Sorry about my previous mis-reading of the code. So you are right, it's not currently implemented. I think ( @ngotelli & @dmcglinn please correct me if I'm wrong) the following procedure is what we had in mind back in Boston discussion:
 - Group plots by treatment. Compute difference in treatment-level metrics. As you said, there would be a single value. 
 -  Shuffle the treatment labels of the plots. 
 - Re-calculate the treatment-level metrics based on the relabeled plots. Repeat.

So even though there is only one number in each iteration, we'd still be able to compare the original numbers to a null expectation to get a p-value. 

I think this procedure has the benefit of being consistent with what we do at the plot-level. However, there may be a caveat of composition - Suppose the two treatments have different species, reshuffling labels would likely inflate the number of species at the gamma-scale for both of them. Would it still be meaningful? For this reason I'm leaning towards Jon's suggestion of jack-knifing. 

What do other folks think?

**Comment by rueuntal** [comment]
Hi @FelixMay - I talked with Brian about this issue and he thinks the relabeling is the way to go. Would you have the time to implement this? If not I'd be happy to add this after our current PRs are merged.

**Comment by FelixMay** [comment]
@rueuntal Well, there are two parts of my answer: Yes, I have time to implement an additional analysis, but not before next week Tuesday. But, I am not convinced yet, that the analysis you described provides the p-value we are looking for.
I my understanding the analysis described above provides a difference **between each group and the overall mean**. But we are looking for a **difference between two groups**, right?
Is that right or do I misunderstand something?
Therefore, I agree to Jon that bootstrapping (i.e. resampling with replacement WITHIN groups might be more relevant for us? Do you agree?
Maybe we should schedule a skype meeting next week and decide how to do it? 
Well, of course, when you have a clear idea how to do it and you do not want to wait until next week, move ahead.

**Comment by rueuntal** [comment]
@FelixMay this looks like something that we'd definitely want to agree on before diving further :) I think both relabeling and bootstrapping can provide a sample (from which to get a p-value) of difference between treatments. In the case of relabeling, we shuffle the labels - compute the difference between the two groups (using new labels) - repeat. These differences between reshuffled treatments can then be compared back with the empirical value. Does that sound right, or am I still missing something?

As for which one is better, I'm not quite so sure. I can see both as having some caveats. One potential issue with relabeling is that if the two treatments have very different composition (e.g., imagine one is from Africa while the other is from North America), reshuffling and recombining plots from different treatments would systematically inflate richness in both treatments (e.g. we may now have both bison and elephant in one treatment), though it doesn't necessarily translate into systematically biased null expectation of the DIFFERENCE. Something similar holds for bootstrapping - by sampling with replacement, in most cases we'd replicate plots within a treatment, thus systematically reduce richness in both treatments. But again, this doesn't necessarily translate into biased estimation of difference.

I think the two approaches are a bit different conceptually - the relabeling gives the envelope of the null expectation (that the two treatments don't differ), while bootstrapping gives a CI around the empirical difference (which can then be compared with zero). Brian prefers relabeling because it's probably more intuitive to the readers (though that may not be right given the communication issue that we have :)). I'm on the fence but like relabeling slightly more because we are already using it to compute p-values at the plot-scale. But I'd be happy to go with bootstrapping if you think that's better. 

If you'd like to talk next week, please let me know!

**Comment by rueuntal** [comment]
The overall p-value for the curves is provided in PR #131 . I will close this thread and open a separate one for the separate issue of p-value at gamma scale.

**Comment by dmcglinn** [comment]
I just discovered this paper by Baddeley et al. 2014 which argue that Loosmore and Ford 's argument that envelopes of the null model cannot be used for significance tests is correct but their global goodness of fit test is overly conservative. 

[baddeleyetal2014.pdf](https://github.com/MoBiodiv/mobr/files/1116969/baddeleyetal2014.pdf)

They argue that the null envelopes are still useful for detecting scale of spatial interaction.  They suggest that we could use a "global envelope test" to detect significance of scale specific patterns. A global envelope is set by the most extreme deviation from the set of null curves and it does not change across scales (flat line). This approach they argue is statistically valid. 

One interesting terminology distinction they made which makes sense is to not refer to the 95% quantiles of the null model as confidence intervals but rather 'acceptance intervals' because observed data outside of the interval suggests the null expectation is unacceptable (i.e., reject Ho). They suggest that confidence intervals apply when one is generating estimates of variance via bootstrap for the observed pattern. Maybe to be even more technically correct we should call the null envelopes "fail to reject intervals" - just kidding ;)
